### PR TITLE
Backport PR #13614 on branch v5.0.x (MNT: Compatibility with Python 3.11)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to manually backport #13614 to v5.0.x branch.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Do the proposed changes need the code-style [fixed](https://docs.astropy.org/en/latest/development/workflow/maintainer_workflow.html#pre-commit_bot)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
@nstarman , the cosmology patch does not seem to apply to v5.0.x, does that sound right to you?

**Comment by pllim** [comment]
Okaaay... it tries to build numpy from source too. I guess we cannot use Stuart's new way for wheels here.

**Comment by pllim** [comment]
Okay, it wasn't scipy or numpy... It was astropy... But why?

https://github.com/astropy/astropy/actions/runs/3054485869/jobs/4926446296

```
      astropy/table/_np_utils.c:196:12: fatal error: longintrepr.h: No such file or directory
        196 |   #include "longintrepr.h"
            |            ^~~~~~~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/gcc' failed with exit code 1
```

**Comment by pllim** [comment]
@eerovaher @nstarman , the patch for pickling isn't compatible with the cosmology code in LTS. How can we fix this? Thanks!

https://github.com/astropy/astropy/actions/runs/3054566720/jobs/4926621567

```
    def generic_recursive_equality_test(a, b, class_history):
        """
        Check if the attributes of a and b are equal. Then,
        check if the attributes of the attributes are equal.
        """
        if PYTHON_LT_3_11:
            dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
        else:
            # NOTE: The call may need to be adapted if other objects implementing a __getstate__
            # with required argument(s) are passed to this function.
            dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__()
        dict_b = b.__dict__
>       for key in dict_a:
E       TypeError: 'NoneType' object is not iterable
```

File: astropy/tests/helper.py [hunk 1]
```diff
@@ -412,11 +413,15 @@ def generic_recursive_equality_test(a, b, class_history):
     Check if the attributes of a and b are equal. Then,
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if PYTHON_LT_3_11:
+        dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    else:
+        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
+        # with required argument(s) are passed to this function.
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__()
```
Comment (dhomeier) [inline astropy/tests/helper.py:421]:
```suggestion
        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
        # with required argument(s) are passed to this function.
        # For a class that has no instance __dict__ and no __slots__, the default state is None
        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
```
The error in cosmology/tests/test_pickle.py was raised by recursion to a function property, which has `__getstate__()` `None`.

File: astropy/tests/helper.py [hunk 2]
```diff
@@ -412,11 +413,16 @@ def generic_recursive_equality_test(a, b, class_history):
...
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if PYTHON_LT_3_11:
+        dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    else:
+        # NOTE: The call may need to be adapted if other objects implementing a __getstate__
+        # with required argument(s) are passed to this function.
+        # For a class that has no instance __dict__ and no __slots__, the default state is None
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
```
Comment (pllim) [inline astropy/tests/helper.py:422]: @dhomeier , do we also need this in `main` and v5.1.x ?

Comment (dhomeier) [reply astropy/tests/helper.py:422]: Yes, we should – I think in principle any function or other thing that returns a state of `None` might be pickled.

Comment (eerovaher) [reply astropy/tests/helper.py:422]: Reading about the Python 3.11 [`object.__getstate__()`](https://docs.python.org/3.11/library/pickle.html#object.__getstate__) I am getting the impression that even the patched code here might not behave as expected in the case where a class uses `__slots__`. Should we patch this code even further?

Comment (dhomeier) [reply astropy/tests/helper.py:422]: Do you know an example of something that implements `__slots__`?
> For a class that has [__slots__](https://docs.python.org/3.11/reference/datamodel.html#object.__slots__) and no instance [__dict__](https://docs.python.org/3.11/library/stdtypes.html#object.__dict__), the default state is a tuple whose first item is None and whose second item is a dictionary mapping slot names to slot values described in the previous bullet.

It _looks_ like we could use `a.__getstate__()[1]` in that case...

Comment (dhomeier) [inline astropy/tests/helper.py:422]:
```suggestion
        # For a class that has no instance __dict__ and no __slots__, the default state is None
        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__() or dict()
        # For a class that has __slots__ and no instance __dict__, the default state is a tuple whose
        # first item is None and whose second item is a dictionary mapping slot names to slot values.
        if insinstance(dict_a, tuple):
            dict_a = dict_a[1]
```

Comment (eerovaher) [reply astropy/tests/helper.py:422]: There are classes in `astropy` that use `__slots__`:
```shell
$ git grep __slots__
CHANGES.rst:4077:  because the class now uses ``__slots__`` any attributes on the class have to
astropy/cosmology/funcs/comparison.py:46:    __slots__ = ("wrapped", )
astropy/table/bst.py:67:    __slots__ = ('val',)
astropy/table/bst.py:108:    __slots__ = ('key', 'data', 'left', 'right')
astropy/table/soco.py:17:    __slots__ = ('key', 'value')
astropy/utils/data_info.py:237:        dct.setdefault('__slots__', [])
astropy/utils/data_info.py:288:    __slots__ = ['_parent_cls', '_parent_ref', '_attrs']
astropy/utils/data_info.py:534:    __slots__ = ['_format_funcs', '_copy_indices']
```

**Comment by Cadair** [comment]
> Okaaay... it tries to build numpy from source too. I guess we cannot use Stuart's new way for wheels here.

If it is compiling numpy then there isn't a wheel for it so the alternative is that it errors.

File: astropy/tests/helper.py (hunk 2 above)
Comment (nstarman) [reply astropy/tests/helper.py:422]: And hopefully more will use slots in py3.10+ with ``dataclass(slots=True)``!

Comment (dhomeier) [reply astropy/tests/helper.py:422]: All those cases also have `__dict__`, so that takes precedence. However there is a weird exception `astropy.utils.data_info`:
```python
from astropy.utils.data_info import DataInfo
DataInfo.__dict__.keys()
dict_keys(['__module__', '__doc__', '_stats', 'attrs_from_parent', 'attr_names', '_attr_defaults', '_attrs_no_copy', '_info_summary_attrs', '__slots__', '_represent_as_dict_attrs', '_construct_from_dict_args', '_represent_as_dict_primary_data', '__init__', '_parent', '__get__', '__set__', '__getstate__', '__setstate__', '_represent_as_dict', '_construct_from_dict', 'info_summary_attributes', 'info_summary_stats', '__call__', '__repr__', '_attrs', '_parent_cls', '_parent_ref', 'description', 'dtype', 'name', 'meta', 'format', 'unit'])

DataInfo.__getstate__(DataInfo)
<member '_attrs' of 'DataInfo' objects>
```
That `member_descriptor` has neither keys nor is it iterable, no idea what use to make of that...

Comment (dhomeier) [reply astropy/tests/helper.py:422]: Could force it somehow to fall back to `a.__dict__` of course

**Comment by nstarman** [comment]
> @eerovaher @nstarman , the patch for pickling isn't compatible with the cosmology code in LTS. How can we fix this? Thanks!

Darn. The file ``astropy/cosmology/tests/test_pickle.py`` has been removed since. I can take a look in a bit.

**Comment by dhomeier** [comment]
> Darn. The file `astropy/cosmology/tests/test_pickle.py` has been removed since. I can take a look in a bit.

It is passing in https://github.com/astropy/astropy/actions/runs/3055240601/jobs/4928078581, isn't it?

**Comment by pllim** [comment]
@dhomeier , this test only exists in LTS branch, not sure why.

@nstarman , any advise on how to fix this properly since you were the main author for that test, if I remember correctly?

**Comment by nstarman** [comment]
Ok, taking a look... Everything's passing! That was easy 😆

File: astropy/tests/helper.py (hunk 2 above)
Comment (dhomeier) [reply astropy/tests/helper.py:422]: > For a class with `__slots__` or no `__dict__`, the default state is not a

Hmm, the Python docs say
"For a class that has **no** instance `__dict__` and **no** `__slots__`, the default state is `None`.".
But in the cosmology example we have
```python
>>> print(FLRW.H0.fvalidate)
<function _validate_to_scalar at 0x126565a80>
>>> hasattr(FLRW.H0.fvalidate, '__slots__')
False
>>> print(FLRW.H0.fvalidate.__dict__)
{}
>>> print(FLRW.H0.fvalidate.__getstate__())
None
```
The above case would still break in the (so far hypothetical) case of something that has no `__dict__`.

**Comment by pllim** [comment]
It does but looks like a7190c4 fixed it? Is that good enough? I see there are still some discussions.

File: astropy/tests/helper.py (hunk 2 above)
Comment (dhomeier) [reply astropy/tests/helper.py:422]: I've modified your suggestion to account for that corner case.
In the sense of "if there is no usable `__getstate__()` return, use `__dict__` that probably does match the pre-3.11 idea better.

**Comment by dhomeier** [comment]
> It does but looks like [a7190c4](https://github.com/astropy/astropy/commit/a7190c45c54eb827513793ec87d779dd5d345c08) fixed it? Is that good enough? I see there are still some discussions.

`generic_recursive_equality_test` does handle the cosmology pickle now; the discussion is just about yet more obscure cases it might need to handle.

**Comment by pllim** [comment]
Okay... Ping me when there is a consensus. Thanks!

File: astropy/tests/helper.py (hunk 2 above)
Comment (dhomeier) [reply astropy/tests/helper.py:422]: @pllim if @eerovaher agrees to the changed suggestion, let's test that and we should be done (and feed it back into 5.1 and main).

Comment (eerovaher) [reply astropy/tests/helper.py:422]: Your edit to my suggestion looks good and if the tests succeed then it would indeed be a good idea to apply that patch to the other branches.

**Comment by pllim** [comment]
`test_biweight_midvariance_masked` failure is unrelated.

**Comment by pllim** [comment]
Do you mind if I squash the commits?

**Comment by eerovaher** [comment]
Good idea to squash, but updates made here to `generic_recursive_equality_test()` should be easily applicable to the other branches after squashing.

**Comment by pllim** [comment]
Okay, squashed. The forward port is at https://github.com/astropy/astropy/pull/13675

**Comment by pllim** [comment]
Huh... @nstarman , do we really need exact match here? Can we use `allclose`?

```
_______________________ TestFlatwCDM.test_toformat_model _______________________

self = <astropy.cosmology.tests.test_flrw.TestFlatwCDM object at 0x7fcf997acb50>
cosmo = FlatwCDM(name="ABCMeta", H0=70 km / (Mpc s), Om0=0.27, w0=-1, Tcmb0=3 K, Neff=3.04, m_nu=[0. 0. 0.] eV, Ob0=None)
to_format = <astropy.cosmology.connect.CosmologyToFormat object at 0x7fcf78113ee0>
method_name = 'Ode'

    def test_toformat_model(self, cosmo, to_format, method_name):
        """Test cosmology -> astropy.model."""
        if method_name is None:  # no test if no method
            return
    
        model = to_format("astropy.model", method=method_name)
        assert isinstance(model, _CosmologyModel)
    
        # Parameters
        expect = tuple([n for n in cosmo.__parameters__ if getattr(cosmo, n) is not None])
        assert model.param_names == expect
    
        # scalar result
        args = np.arange(model.n_inputs) + 1
    
        got = model.evaluate(*args)
        expected = getattr(cosmo, method_name)(*args)
        assert np.all(got == expected)
    
        got = model(*args)
        expected = getattr(cosmo, method_name)(*args)
>       assert np.all(got == expected)
E       assert False
E        +  where False = <function all at 0x7fcfaf7971c0>(0.2523877726514899 == 0.25238777265148993)
E        +    where <function all at 0x7fcfaf7971c0> = np.all

.../cosmology/io/tests/test_model.py:119: AssertionError
```

**Comment by pllim** [comment]
CI finally green again. Thank you, all!
